ci: build universal release on macOS

carlocab opened the pull request:
I haven't tested this, but this implements one of my suggestions from #13880.

Closes #13880.

carlocab commented:
Oops, the `publish` step needs updating too.

carlocab commented on .github/workflows/release.yml:
```diff
@@ -118,6 +118,63 @@ jobs:
+          mv -f "$libintl_basename" "${gettext_prefix}/lib/${libintl_basename}"
+      - name: Build release
+        run: |
+          CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX:PATH= -DCMAKE_OSX_DEPLOYMENT_TARGET=11 -DCMAKE_OSX_ARCHITECTURES=\\\"arm64;x86_64\\\""
```
Tried testing this in my fork. Unfortunately, `make` chokes on the `;` inside `CMAKE_OSX_ARCHITECTURES`. Not sure how to get around that.

justinmk replied:
Try `\;`

(in cmake, `;` is how a "list" is internally encoded, but `\;` should work around this)

carlocab replied:
Yes, that seems to work. Thanks.

carlocab commented:
Hmmm, it looks like some of the dependent builds also need adjusting for this to work.

justinmk commented on .github/workflows/notes.md:
```diff
@@ -18,12 +18,18 @@ ${NVIM_VERSION}
 2. Extract: `tar xzvf nvim-macos.tar.gz`
 3. Run `./nvim-osx64/bin/nvim`
 
+### macOS 11+ (Universal)
```
do we need both? or can we have only the universal build?

clason replied:
Big +1 for just providing a single universal build.

carlocab replied:
Having a single universal build is tricky. Ideally you'll want to set different `MACOSX_DEPLOYMENT_TARGET`/`CMAKE_OSX_DEPLOYMENT_TARGET`s for each arch slice (`10.11` for `x86_64`, `11` for `arm64`), and `cmake` doesn't make it easy to set arch-dependent depoloyment targets. Setting only one means you end up with a binary with restricted portability (deployment target too high) or with a possibly broken `arm64` slice (deployment target too low).

One way to do this is to build for each arch separately and then use `lipo` to bundle them all up.

If I can get this approach for a universal build working then this alternative approach (build separately + bundle together with `lipo`) should be relatively straightforward.

carlocab replied:
> If I can get this approach for a universal build working then this alternative approach (build separately + bundle together with `lipo`) should be relatively straightforward.

I lied; sorry. Under a default build and in the absence of the ability to run ARM64 binaries on an X64 machine, this is pretty much impossible. This is because we pre-compile Lua sources into byte-code, which requires a `lua[jit]` compiler that runs on the host machine.

carlocab replied:
Works for me. Though some initial testing suggests that we need not make that tradeoff.

This hinges, however, on whether the byte-code compiled with an x64 luajit and embedded into the (universal) nvim binary will work when executed on an arm64 machine. Would you know? I was actually already contemplating on tagging you here to ask this, so it's convenient you've dropped by.

gpanders replied:
According to https://luajit.org/extensions.html#string_dump:

>The generated bytecode is portable and can be loaded on any architecture that LuaJIT supports, independent of word size or endianess. However the bytecode compatibility versions must match. Bytecode stays compatible for dot releases (x.y.0 → x.y.1), but may change with major or minor releases (2.0 → 2.1) or between any beta release.

So it sounds like it should work.

justinmk commented:
nice, thank you! your notes in https://github.com/neovim/neovim/issues/13880#issuecomment-1160187655 would be valuable to include in the commit message.

carlocab commented:
Ok, I'm blocked on the dependent builds. I've managed to get a few of them to work with the tweaks I've made here:
```
❯ file .deps/usr/{bin,lib}/* | rg Mach
.deps/usr/bin/unterm:                  Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit x86_64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|PIE>] [arm64:Mach-O 64-bit arm64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|PIE>]
.deps/usr/bin/vterm-ctrl:              Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit x86_64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|PIE>] [arm64:Mach-O 64-bit arm64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|PIE>]
.deps/usr/bin/vterm-dump:              Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit x86_64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|PIE>] [arm64:Mach-O 64-bit arm64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|PIE>]
.deps/usr/lib/libmsgpackc.2.0.0.dylib: Mach-O 64-bit arm64 dynamically linked shared library, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|NO_REEXPORTED_DYLIBS>
.deps/usr/lib/libunibilium.a:          Mach-O universal binary with 2 architectures: [x86_64:current ar archive random library] [arm64:current ar archive random library]
.deps/usr/lib/libuv.a:                 Mach-O universal binary with 2 architectures: [x86_64:current ar archive random library] [arm64]
.deps/usr/lib/libvterm.a:              Mach-O universal binary with 2 architectures: [x86_64:current ar archive random library] [arm64:current ar archive random library]
```
Most of these come out as universal binaries. Not sure why `msgpack` is compiling only as `arm64`. I'll take a closer look at that.

LuaJIT isn't cooperative, though, which is a bit of a show-stopper.

clason commented:
That is quite possible, and would explain why, e.g., Julia doesn't provide universal binaries. In this case, I'd say it's fine to have fully separate releases (with ARM only available for macOS 11+).

carlocab commented:
Ok, that was a bit more work than I expected, but it works: https://github.com/carlocab/neovim/releases/download/nightly/nvim-macos-universal.tar.gz

On ARM64, you will need to do
```
install_name_tool -change "@@HOMEBREW_PREFIX@@/opt/gettext/lib/libintl.8.dylib" "@executable_path/../libs/libintl.8.dylib" ./nvim-universal/bin/nvim
```
to get a working binary for now since I messed up the install-name rewriting somewhere.

However, once doing that,
```
./nvim-universal/bin/nvim -i NONE -u NONE -c checkhealth
```
and
```
./nvim-universal/bin/nvim -c checkhealth
```
both seem to work with no problems.

If someone wants to help me test this build more thoroughly, that'd be much appreciated.

There's still a bit of work left here. I'll write up what I think still needs looking at soon, but since this is working now I'm marking this as ready for review.

carlocab commented on third-party/cmake/BuildLuajit.cmake:
```diff
@@ -51,30 +52,58 @@ if(CMAKE_SYSTEM_NAME MATCHES "OpenBSD")
+    set(STATIC_CC "${LJ_C_COMPILER} -arch ${ARCH}")
+    set(DYNAMIC_CC "${LJ_C_COMPILER} -arch ${ARCH} -fPIC")
+    set(TARGET_LD "${LJ_C_COMPILER} -arch ${ARCH}")
+    list(APPEND LUAJIT_THIN_EXECUTABLE "${LJ_SRC_DIR}-${ARCH}/src/luajit")
+    list(APPEND LUAJIT_THIN_STATIC_LIB "${LJ_SRC_DIR}-${ARCH}/src/libluajit.a")
+    list(APPEND LUAJIT_THIN_DYN_LIB "${LJ_SRC_DIR}-${ARCH}/src/libluajit.so")
```
I might be relying on CMake implementation details here.

carlocab commented:
Ok, things to do/verify:

- [ ] style, variable names, commit messages -- I paid very little attention to this, so this is probably a bit of a mess at the moment.
- [ ] Check that `codesign`ing the binaries works. This is needed to run the built binaries on Apple Silicon.
- [ ] Fix LuaJIT build. I still get intermittent build failures here, so there's probably something I don't grok about `cmake` that I need to fix. (Help appreciated.)
- [ ] Is LuaJIT bytecode portable? If not, we probably need to skip byte-compiling the Lua sources in order to have a working universal binary. Or find a way to cross-compile the bytecode.
- [ ] Single universal build (instead of two builds: a universal one and an Intel-only one)

carlocab commented on third-party/cmake/BuildLuajit.cmake:
```diff
@@ -30,13 +30,14 @@ function(BuildLuajit)
     add_custom_command(
       TARGET ${_luajit_TARGET}
-      COMMAND ${CMAKE_COMMAND} -E create_symlink luajit-2.1.0-beta3 ${DEPS_BIN_DIR}/luajit)
+      COMMAND ${CMAKE_COMMAND} -E create_symlink luajit-2.1.0-beta3 ${DEPS_BIN_DIR}/${_luajit_TARGET})
```
This seems to make my build woes go away. I don't understand why, but it does.

dundargoc replied:
lolwat? Aren't they the same string?

carlocab replied:
Not exactly. I build intermediate luajit targets [here](https://github.com/neovim/neovim/blob/fd2329a80be0ff405e058cbed0d5ebc4201e3a73/third-party/cmake/BuildLuajit.cmake#L91-L96):
```cmake
    BuildLuaJit(TARGET "luajit-${ARCH}"
        BUILD_COMMAND ${BUILDCMD_UNIX}
        CC=${LUAJIT_C_COMPILER} STATIC_CC=${STATIC_CC}
        DYNAMIC_CC=${DYNAMIC_CC} TARGET_LD=${TARGET_LD}
        PREFIX=${DEPS_INSTALL_DIR}
        ${DEPLOYMENT_TARGET})
```

carlocab commented on third-party/cmake/BuildLuajit.cmake:
```diff
@@ -51,30 +52,58 @@ if(CMAKE_SYSTEM_NAME MATCHES "OpenBSD")
+    list(APPEND LUAJIT_THIN_DYLIBS "${LUAJIT_SRC_DIR}-${ARCH}/src/libluajit.so")
+    list(APPEND LUAJIT_THIN_DEPS "luajit-${ARCH}")
+
+    BuildLuaJit(TARGET "luajit-${ARCH}"
+        BUILD_COMMAND ${BUILDCMD_UNIX}
+        CC=${LUAJIT_C_COMPILER} STATIC_CC=${STATIC_CC}
+        DYNAMIC_CC=${DYNAMIC_CC} TARGET_LD=${TARGET_LD}
+        PREFIX=${DEPS_INSTALL_DIR}
+        ${DEPLOYMENT_TARGET})
```
For reference (I should/will add this to a commit message):

The extra flags are taken from https://luajit.org/install.html, in the section on cross-compiling.

We could technically define a `CROSS` variable instead, but this would require creating executable stubs like `arm64-apple-darwin-clang` that basically just call `clang -arch arm64`.

dundargoc reviewed (approved):
Took a look at it. The cmake part seems fine but I can't confirm as I don't own a mac. The release.yml could use a matrix (if possible) to avoid duplication as you said.

dundargoc commented:
Hmm, not sure if/why we need so much build logic outside of cmake (not just for the universal but the original macos job as well). Feels like we should move the logic into cmake.

Note: this was an open comment for future consideration/PR, this isn't a review comment for this PR.

carlocab commented:
Some documentation on byte code portability: https://luajit.org/running.html#opt_b

justinmk commented:
> Single universal build (instead of two builds: a universal one and an Intel-only one) -- how safe is setting a deployment target less than 11 on Apple Silicon?

If it helps reach the goal of "one download for macos", it's fine to make macos 11 the minimum.

carlocab commented:
Ok, I'm ready to clean this PR up and finish it off. I intend to provide only a single universal build instead of two separate ones as originally proposed.

I have one question I'd appreciate an answer on before proceeding. Is there any reason why we want to link dynamically to `libintl`? I can probably adjust the workflow to make sure it's linked statically instead if desired.

I suspect static linkage is preferable, but perhaps I'm missing something here.

justinmk commented:
> there any reason why we want to link dynamically to `libintl`?

The current/old approach to preparing a macos download was simply a desperate attempt to get it to work.  There's no other reason for any of the decisions there beyond that :)

> I can probably adjust the workflow to make sure it's linked statically instead if desired.

💯 I think this is nicer for users, than having a carefully arrange bag of libraries in a folder. Opens the door for a "portal binary" too?

clason commented:
Re `MACOSX_DEPLOYMENT_TARGET`: LibUV only supports [macOS 10.15+](https://github.com/libuv/libuv/blob/v1.x/SUPPORTED_PLATFORMS.md#supported-platforms), so it is not reasonable to promise a working build below that; hence I suggest raising the deployment target accordingly.

If it makes life easier, I'd even suggest raising to macOS 11.0 across the bar for our releases?

carlocab commented:
Ok, that was a lot more finicky than I expected (what else is new), but here's a statically linked binary you can test out: https://github.com/carlocab/neovim/actions/runs/2543816551

I broke the generation of version information, but that should be easy to fix (for real this time, I hope).

justinmk commented:
> here's a statically linked binary you can test out: https://github.com/carlocab/neovim/actions/runs/2543816551

works for me on 12.4. Wonder if we should rename the inner folder from nvim-osx64 to nvim-macos?

clason commented:
Also, remove the old macOS release in favor of the universal one?

justinmk commented on third-party/cmake/BuildLuajit.cmake:
```diff
@@ -51,30 +52,58 @@ if(CMAKE_SYSTEM_NAME MATCHES "OpenBSD")
+elseif(CMAKE_OSX_ARCHITECTURES AND APPLE)
+
+  set(LUAJIT_SRC_DIR "${DEPS_BUILD_DIR}/src/luajit")
+  foreach(ARCH IN LISTS CMAKE_OSX_ARCHITECTURES)
```
```suggestion
  # Cross-compile LuaJIT for our macos "universal" binary. #19029
  foreach(ARCH IN LISTS CMAKE_OSX_ARCHITECTURES)
```

justinmk commented on third-party/cmake/BuildLuajit.cmake:
```diff
@@ -51,30 +52,58 @@ if(CMAKE_SYSTEM_NAME MATCHES "OpenBSD")
-    ${DEPLOYMENT_TARGET})
+    ${DEPLOYMENT_TARGET} install)
+
+elseif(CMAKE_OSX_ARCHITECTURES AND APPLE)
```
should probably be an error of CMAKE_OSX_ARCHITECTURES is defined and APPLE is not?

carlocab replied:
The CMake documentation states that `CMAKE_OSX_ARCHITECTURES` is ignored on inapplicable platforms, and the intention of this check is to follow that behaviour.

That said, I don't think this works as intended, because I think you end up not building LuaJIT if you define `CMAKE_OSX_ARCHITECTURES` on Linux. Will need to fix that.

jamessan commented on .github/workflows/release.yml:
```diff
@@ -118,6 +118,65 @@ jobs:
+              install_name_tool -change "$lib" "@executable_path/../$relname" nvim-universal/bin/nvim
+            fi
+          done
+          codesign --force --sign - nvim-universal/bin/nvim
```
Don't we need an Apple developer key or something for this?

carlocab replied:
Nope; that's why we're using ad ad-hoc signature (`--sign -`). That said, this line goes away when I fix the build to link statically to `libintl`, but only because `clang` executes something similar to this under the hood.

clason replied:
Another strong argument for static linking; let's just go with that!

carlocab commented:
Marking as draft while I'm cleaning up.

carlocab committed:
build(luajit): support universal builds on macOS

To build universal binaries on macOS, one typically only needs to pass
multiple `-arch` flags to `clang`.

Unfortunately, this strategy causes LuaJIT builds to fail. To work
around this, we build LuaJIT for each requested architecture
individually first and then use `lipo` to package each architecture
slice into a universal binary.

To be able to do this on an Intel macOS host or an M1 macOS host without
a Rosetta installation, we need some special flags in order to tell
LuaJIT that it is cross-compiling for a different target. See [1] for
details.

[1] https://luajit.org/install.html#cross

carlocab committed:
build(deps): support universal builds on macOS

CMake can handle building universal binaries on macOS using the
`CMAKE_OSX_ARCHITECTURES` variable. Let's pass this variable to the
relevant dep builds. We use the `LIST_SEPARATOR` argument to prevent the
shell from interpreting the `;` that CMake uses as a list separator.

For dependencies that don't build using CMake, we only need to make sure
that the compiler is invoked with the correct `-arch` flags. The
compiler does the rest.

The only exception to this is the LuaJIT build, which we handle
separately as a special case.

carlocab committed:
ci(release): build a universal binary on macOS

After some tweaks to our dep builds, we can now build a universal binary
for macOS by using `CMAKE_OSX_ARCHITECTURES`. So, let's do that. This
requires a number of additional changes:

1. We need to build on macOS 11, since earlier versions do not support
   building universal (M1 + Intel) binaries.
2. We need to provision a universal `libintl`. The linker will look for
   an ARM64 version of this library when linking the `nvim` binary.

While we're here:
1. Link statically to `libintl`. This allows to to avoid having to do
   any install name rewriting or codesigning to package Neovim.
2. Bump the `MACOSX_DEPLOYMENT_TARGET` to `11`. We're already using a
   `libintl` built by Homebrew (through the pre-installed version of
   `gettext`), and that is built for macOS 11.

In order to ensure we link to `libintl.a` instead of `libintl.dylib`, we
have to make sure that CMake can't find the latter. This ideally should
be a matter of doing `brew unlink gettext`. However, CMake is too adept
at finding things that Homebrew has installed (even when not linked), so
we have to do a bit more than that. This appears in the additional step
ensuring static linkage to `libintl`.

We end up breaking some Homebrew-installed software in the process, and
some of these software is called during our build (e.g. curl, git,
wget). To avoid any adverse effects, let's just uninstall them.

carlocab commented:
Ok, I think this is ready.

Test release at https://github.com/carlocab/neovim/releases/tag/nightly.

carlocab commented:
Some sanity checks:
```
❯ file nvim-macos/{bin/nvim,**/*.so}
nvim-macos/bin/nvim:             Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit x86_64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|PIE|HAS_TLV_DESCRIPTORS>] [arm64]
nvim-macos/lib/nvim/parser/c.so: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit x86_64 bundle, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL>] [arm64:Mach-O 64-bit arm64 bundle, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL>]
❯ otool -L nvim-macos/{bin/nvim,**/*.so}
nvim-macos/bin/nvim:
        /System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 1141.1.0)
        /usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.0.0)
        /usr/lib/libutil.dylib (compatibility version 1.0.0, current version 1.0.0)
nvim-macos/lib/nvim/parser/c.so:
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.0.0)
```
And the build successfully makes use of the installed `gettext` even after we mangled it for static linkage: https://github.com/carlocab/neovim/runs/7044719516?check_suite_focus=true#step:8:1920
```
-- Found Gettext: /usr/local/bin/msgmerge (found version "0.21") 
```

carlocab commented:
> Nice! Works well on Intel (even without complaining about an unknown developer) and M1 (where it still does complain...)!

I think you can get around this by doing
```bash
codesign -f -s - nvim-macos/{bin/nvim,**/*.so}
```
Let me know if that helps. If so, we could consider adding it to the release notes as an extra step for M1 macOS.

clason commented:
> I think you can get around this by doing
> 
> ```shell
> codesign -f -s - nvim-macos/{bin/nvim,**/*.so}
> ```

`bin/nvim: Operation not permitted`

> Let me know if that helps. If so, we could consider adding it to the release notes as an extra step for M1 macOS. This may require an existing developer tools (CLT/Xcode) installation, though.

I think it's easier to point people to the Privacy & Security preference pane?

> I'm not able to reproduce the complaint about an unknown developer locally, even on my M1 mac, though, so 🤷. I've only tried accessing it over SSH though, but I don't see why that should make a difference.

That's weird. There must be some setting that controls this which I forgot to toggle?

carlocab commented:
> > I think you can get around this by doing
> > ```shell
> > codesign -f -s - nvim-macos/{bin/nvim,**/*.so}
> > ```
> 
> `bin/nvim: Operation not permitted`
> 

That command works for me, so 🤷.

> > Let me know if that helps. If so, we could consider adding it to the release notes as an extra step for M1 macOS. This may require an existing developer tools (CLT/Xcode) installation, though.
> 
> I think it's easier to point people to the Privacy & Security preference pane?
> 

Yea, maybe. I don't know. That sounds a bit verbose for the release notes for me though. It can be added to some documentation somewhere else unless it's already there.

> That's weird. There must be some setting that controls this which I forgot to toggle? I can't reproduce it on my Intel mac, either; I assumed that this just came with less paranoia due to lack of M1's security enclave, but that seems not to be it...

Maybe. Apple definitely has tighter security on code signatures and whatnot on M1, so it showing up on your M1 Mac but not your Intel one isn't too surprising, though.

clason commented:
> That command works for me, so 🤷.

Yeah, sorry, fish issue. It works, but then the error when opening is "nvim cannot be opened because Apple cannot check it for malicious software".

> Maybe. Apple definitely has tighter security on code signatures and whatnot on M1, so it showing up on your M1 Mac but not your Intel one isn't too surprising, though.

Makes sense.

carlocab commented:
Just to test our `CMAKE_OSX_ARCHITECTURES` handling, we can try a build with some extra architectures:
```
❯ file nvim-macos/{bin/nvim,**/*.so}
nvim-macos/bin/nvim:             Mach-O universal binary with 4 architectures: [x86_64:Mach-O 64-bit x86_64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|PIE|HAS_TLV_DESCRIPTORS>] [x86_64_haswell subarchitecture=8] [arm64] [arm64e (caps: 0x2)]
nvim-macos/lib/nvim/parser/c.so: Mach-O universal binary with 4 architectures: [x86_64:Mach-O 64-bit x86_64 bundle, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL>] [x86_64_haswell subarchitecture=8:Mach-O 64-bit x86_64_haswell subarchitecture=8 bundle, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL>] [arm64:Mach-O 64-bit arm64 bundle, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL>] [arm64e (caps: 0x2)]
```
See the `nvim-macos-12` artefact at https://github.com/carlocab/neovim/actions/runs/2557307648.

-----

> > there any reason why we want to link dynamically to `libintl`?
> 
> The current/old approach to preparing a macos download was simply a desperate attempt to get it to work. There's no other reason for any of the decisions there beyond that :)
> 

Oh dear. I should've worked on this sooner, then. Feel free to ping me for build-related macOS issues in the future. I'm not an expert, but I know some of the ways around things.

> > I can probably adjust the workflow to make sure it's linked statically instead if desired.
> 
> 💯 I think this is nicer for users, than having a carefully arranged bag of libraries in a folder. Opens the door for a "portable binary" too?

This is done. What do you mean by a "portable binary", though? The existing releases seem pretty portable to me, and we improve that a little by linking `libintl` statically.

Do you mean you want an `nvim` binary that I can place wherever I like? That might require embedding the runtime into the binary, which is probably doable.

> Your improvements here are extremely appreciated!

> Still, big step forward, thank you for this effort! (Stellar commit messages, too!)

❤️

clason commented:
@carlocab that broke building from source for me:

```
~/b/neovim ❯❯❯ env CMAKE_BUILD_TYPE=Release make install                                           (master|✓)
cd .deps && \
		cmake -G 'Ninja'   \
		 /Users/clason/build/neovim//third-party
-- Found GNU Make at /usr/bin/make
-- Configuring done
-- Generating done
CMake Error:
  Running

   '/usr/local/bin/ninja' '-C' '/Users/clason/build/neovim/.deps' '-t' 'recompact'

  failed with:

   ninja: error: build.ninja:627: expected '=', got lexing error
   install && /usr/local/Cellar/cmake/3.23.2/bin/cmake -E touch /Users/cla...
           ^ near here



CMake Generate step failed.  Build files cannot be regenerated correctly.
make: *** [build/.ran-third-party-cmake] Error 1
```

clason commented:
Seems to me that it's missing a default fallback `CMAKE_OSX_ARCHITECTURES`, like we have for `MACOSX_DEPLOYMENT_TARGET`?

clason commented:
Hmmm, that's not enough, it seems:
```
~/b/neovim ❯❯❯ env CMAKE_OSX_ARCHITECTURES=x86_64 make                                             (master|✓)
cd .deps && \
		cmake -G 'Ninja'   \
		 /Users/clason/build/neovim//third-party
-- Found GNU Make at /usr/bin/make
-- Configuring done
-- Generating done
CMake Error:
  Running

   '/usr/local/bin/ninja' '-C' '/Users/clason/build/neovim/.deps' '-t' 'recompact'

  failed with:

   ninja: error: build.ninja:611: expected variable name
   && /usr/local/Cellar/cmake/3.23.2/bin/cmake -E touch /Users/clason/buil...
   ^ near here



CMake Generate step failed.  Build files cannot be regenerated correctly.
make: *** [build/.ran-third-party-cmake] Error 1
```
